app/services/work_package_bulk_update_service.rb

```diff
@@ -0,0 +1,157 @@
+    prepare(params)
+    unsaved_work_package_ids = []
+    moved_work_packages      = []
+    @work_packages.sort!
+
+    @work_packages.each do |work_package|
+      work_package.reload
+
+      if copy
+        Redmine::Hook.call_hook(:controller_work_packages_move_before_save,
+                                params:         params,
+                                work_package:   work_package,
+                                target_project: @target_project,
+                                copy:           !!@copy)
+
+        if r = work_package.move_to_project(@target_project,
+                                            @new_type,  copy:         @copy,
+                                                        attributes:   permitted_params(params),
+                                                        journal_note: @notes)
+          moved_work_packages << r
+        else
+          unsaved_work_package_ids << work_package.id
+        end
+
+      else
+        work_package.add_journal(User.current, @notes)
+
+        # filter parameters by whitelist and add defaults
+        attributes = parse_params_for_bulk_work_package_attributes params, work_package.project
+        work_package.assign_attributes attributes
+
+        Redmine::Hook.call_hook(:controller_work_package_bulk_before_save,
+                                params: params, work_package: work_package)
+        JournalObserver.instance.send_notification = params[:send_notification] == '0' ? false : true
+        unless work_package.save
+          unsaved_work_package_ids << work_package.id
+        end
+
+      end
+    end
+
+    { moved_work_packages:      moved_work_packages,
+      unsaved_work_package_ids: unsaved_work_package_ids,
+      copy: @copy }
+  end
+
+  private
+
+  def prepare(params)
+    @work_packages.sort!
+    @copy             = params.has_key? :copy
+    @allowed_projects = WorkPackage.allowed_target_projects_on_move
+    @target_project   = @allowed_projects.detect { |p| p.id.to_s == params[:new_project_id].to_s } if params[:new_project_id]
+    @target_project   ||= @project
+    @new_type         = params[:new_type_id].blank? ? nil : @target_project.types.find_by_id(params[:new_type_id])
+    @notes            = params[:notes]
+    @notes            ||= ''
+
+    @custom_fields      = @projects.map(&:all_work_package_custom_fields)
+                          .inject { |memo, c| memo & c }
+    @assignables        = @projects.map(&:possible_assignees).inject { |memo, a| memo & a }
+    @responsibles       = @projects.map(&:possible_responsibles).inject { |memo, a| memo & a }
+    @available_statuses = @projects.map { |p| Workflow.available_statuses(p) }.inject { |memo, w| memo & w }
+
+    if params.has_key? :copy
+      @types = @target_project.types
+    else
+      @types = @projects.map(&:types).inject { |memo, t| memo & t }
+    end
+    self
+  end
+
+  def parse_params_for_bulk_work_package_attributes(params, project)
```

Comment:
Assignment Branch Condition size for parse_params_for_bulk_work_package_attributes is too high. [20.83/15]